src/common/macro/src/admin_fn.rs

```diff
@@ -0,0 +1,236 @@
+fn extract_handler_type(arg_types: &[Type]) -> Result<&Ident, syn::Error> {
+    match &arg_types[0] {
+        Type::Reference(TypeReference { elem, .. }) => match &**elem {
+            Type::Path(TypePath { path, .. }) => Ok(&path
+                .segments
+                .first()
+                .expect("Expected a reference of handler")
+                .ident),
+            other => {
+                error!(other.span(), "Expected a reference of handler")
+            }
+        },
+        other => {
+            error!(other.span(), "Expected a reference of handler")
+        }
+    }
+}
+
+/// Build the function struct
+#[allow(clippy::too_many_arguments)]
+fn build_struct(
+    attrs: Vec<Attribute>,
+    vis: Visibility,
+    fn_name: &Ident,
+    name: Ident,
+    sig_fn: Ident,
+    ret: Ident,
+    handler_type: &Ident,
+    display_name_ident: Ident,
+) -> TokenStream {
+    let display_name = display_name_ident.to_string();
+    let ret = Ident::new(&format!("{ret}_datatype"), ret.span());
+    let uppcase_display_name = display_name.to_uppercase();
+    // Get the handler name in function state by the argument ident
+    let (handler, snafu_type) = match handler_type.to_string().as_str() {
+        "ProcedureServiceHandlerRef" => (
+            Ident::new("procedure_service_handler", handler_type.span()),
+            Ident::new("MissingProcedureServiceHandlerSnafu", handler_type.span()),
+        ),
+
+        "TableMutationHandlerRef" => (
+            Ident::new("table_mutation_handler", handler_type.span()),
+            Ident::new("MissingTableMutationHandlerSnafu", handler_type.span()),
+        ),
+        handler => ok!(error!(
+            handler_type.span(),
+            format!("Unknown handler type: {handler}")
+        )),
+    };
+
+    quote! {
+        #(#attrs)*
+        #[derive(Debug)]
+        #vis struct #name;
+
+        impl fmt::Display for #name {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                write!(f, #uppcase_display_name)
+            }
+        }
+
+
+        impl Function for #name {
+            fn name(&self) -> &'static str {
+                #display_name
+            }
+
+            fn return_type(&self, _input_types: &[ConcreteDataType]) -> Result<ConcreteDataType> {
+                Ok(ConcreteDataType::#ret())
+            }
+
+            fn signature(&self) -> Signature {
+                #sig_fn()
+            }
+
+            fn eval(&self, func_ctx: FunctionContext, columns: &[VectorRef]) ->  Result<VectorRef> {
+                // Ensure under the `greptime` catalog for security
+                ensure_greptime!(func_ctx);
+
+                let columns_num = columns.len();
+                let rows_num = if columns.is_empty() {
+                    1
+                } else {
+                    columns[0].len()
+                };
+                let columns = Vec::from(columns);
+
+
+                std::thread::spawn(move || {
```

Comment:
add a comment to this `thread::spawn`. The original comment is not moved to here.